src/shim/runtime/v2/rune/v2/agent.go

```diff
@@ -0,0 +1,279 @@
+	if err != nil {
+		return nil, err
+	}
+
+	// Create dir for store unionfs image (based on sefs)
+	sefsDir := filepath.Join(c.Bundle, "rootfs/images/", cid, "sefs")
+	lowerDir := filepath.Join(sefsDir, "lower")
+	upperDir := filepath.Join(sefsDir, "upper")
+	for _, dir := range []string{lowerDir, upperDir} {
+		if err := os.MkdirAll(dir, 0700); err != nil && !os.IsExist(err) {
+			return nil, err
+		}
+	}
+
+	r := &grpc.PullImageRequest{
+		Image:       req.Image,
+		ContainerId: cid,
+	}
+	resp, err := c.sendReq(ctx, r)
+	if err != nil {
+		c.Logger().Errorf("agent enclave container pull image err. %v", err)
+		return nil, err
+	}
+	response := resp.(*grpc.PullImageResponse)
+	return &image.PullImageResp{
+		ImageRef: response.ImageRef,
+	}, nil
+}
+
+// The function creates agent enclave container based on a pre-installed OCI bundle
+func createAgentContainer(ctx context.Context, s *service, r *taskAPI.CreateTaskRequest) (*runc.Container, error) {
+	dir := filepath.Join(agentContainerRootDir, r.ID)
+	upperDir := path.Join(dir, "upper")
+	workDir := path.Join(dir, "work")
+	destDir := path.Join(dir, "merged")
+	for _, dir := range []string{upperDir, workDir, destDir} {
+		if err := os.MkdirAll(dir, 0700); err != nil && !os.IsExist(err) {
+			return nil, err
+		}
+	}
+
+	var options []string
+	// Set index=off when mount overlayfs
+	options = append(options, "index=off")
+	options = append(options,
+		fmt.Sprintf("lowerdir=%s", filepath.Join(agentContainerPath, "rootfs")),
+		fmt.Sprintf("workdir=%s", workDir),
+		fmt.Sprintf("upperdir=%s", upperDir),
+	)
+	r.Rootfs = append(r.Rootfs, &types.Mount{
+		Type:    "overlay",
+		Source:  "overlay",
+		Options: options,
+	})
+	r.Bundle = destDir
+
+	fs.CopyFile(filepath.Join(r.Bundle, configFilename), filepath.Join(agentContainerPath, configFilename))
+
+	// Create Stdout and Stderr file for agent enclave container
+	r.Stdout = filepath.Join(agentContainerRootDir, r.ID, r.ID+"-stdout")
+	r.Stderr = filepath.Join(agentContainerRootDir, r.ID, r.ID+"-stderr")
+	for _, file := range []string{r.Stdout, r.Stderr} {
+		f, err := os.Create(file)
+		if err != nil {
+			return nil, err
+		}
+		defer f.Close()
+	}
+
+	agentContainer, err := runc.NewContainer(ctx, s.platform, r)
+	if err != nil {
+		return nil, err
+	}
+
+	return agentContainer, nil
+}
+
+// Cleanup the agent enclave container resource
+func cleanupAgentContainer(ctx context.Context, id string) error {
+	var cfg shimconfig.Config
+	if _, err := toml.DecodeFile(constants.ConfigurationPath, &cfg); err != nil {
+		return err
+	}
+	rootdir := cfg.Containerd.AgentContainerRootDir
+	path := filepath.Join(rootdir, id, "merged")
+
+	ns, err := namespaces.NamespaceRequired(ctx)
+	if err != nil {
+		return err
+	}
+	runtime, err := runc.ReadRuntime(path)
+	if err != nil {
+		return err
+	}
+	opts, err := runc.ReadOptions(path)
+	if err != nil {
+		return err
+	}
+	root := process.RuncRoot
+	if opts != nil && opts.Root != "" {
+		root = opts.Root
+	}
```

Comment:
Nit: This code is a bit "dense". I'd consider a blank line between logical blocks to make it more readable.